### DISCUSSION TITLE
Default behavior for build profile given in the command-line

### DISCUSSION THREAD

**Discussion by mosteo** [alire-project/alire#1140] [Polls]
Related issue: https://github.com/alire-project/alire/pull/1119

Currently, by default Alire will build the root crate in dev mode, and dependencies in release mode. (This is a safe release mode, in the sense that checks are not omitted. It boils down to compiling with `-O3` and inlining.)

While discussing the addition of switches to use the same profile in the root and dependencies, a debate about expectations arose:

When doing `alr build` with an explicit profile switch, do you expect it to apply to the root crate only or to all dependencies too?

E.g.: `alr build --development` will currently only build the root crate in dev mode, with deps in release mode.

Please answer this poll to better understand your expectations, in the sense of what would you expect if you didn't know what I just explained.

**Comment by OCTAGRAM** [comment]
Is symbolic stacktrace available in release dependencies?

**Reply by mosteo** [reply]
I don't think so since `-g` is not present in the compiler switches in release mode. The binder switches in the project generated by default do include `-Es`, which is for [symbolic tracebacks](https://gcc.gnu.org/onlinedocs/gnat_ugn/Symbolic-Traceback.html), but that doesn't seem to imply `-g`.

**Comment by fedor-rybin** [comment]
I would say all, or better yet give an option for interactive per-component selection

**Comment by stcarrez** [comment]
+1 to apply on all crates through dependencies.

In development mode, I expect to be able to use a debugger to trouble shot an issue (ok, it is rare but it happens).
If dependent crates are not compiled with the same options (`-g -O0` for proper debugging), this makes debugging with gdb difficult.  Note that compiling a dependent crate with `-g -O2` will not help since it breaks almost all dump of local variables by gdb.

The issue with `-O3` and inlining is that breakpoints will not work when you try to put the breakpoint within that inlined function.

**Reply by Joebeazelman** [reply]
Perhaps, I'm clueless about this issue,but does it have to be all or nothing? How about adding additional profiles to differentiate how the build is applied?  If you want to apply the settings to the root only, for example, you would use **--development** profile. On the other hand, if  you want them to apply to dependencies as well, you would use **--development-with-dependencies** profile.

Alternatively, you can set an additional environment or command-line modifier flag, **--root-only**, or **--apply-to-dependencies**. This might be a better choice as it is factored out, reducing the number of profiles.

**Reply by mosteo** [reply]
Yes, we are going in the direction of keeping the current switches and behavior and adding some extra switch. This discussion was more to gauge what uninformed people would expect. Fortunately, all of us technical people enjoy reading the docs, right? 

(Seriously, I do.)

**Comment by Fabien-Chouteau** [comment]
`All crates being compiled through dependencies. `
Note that this behavior will be very annoying for the `--validation` mode. The build will fail for any style check or warning in any of your dependencies. It a no-go for me.

**Reply by stcarrez** [reply]
It's not a good idea to propagate style checks to dependencies because each library could have their own settings.

**Reply by Fabien-Chouteau** [reply]
Yeah we are not doing that and we won't do it.
